rust/src/datatypes/schema.rs

```diff
@@ -0,0 +1,475 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Schema
+
+use std::{
+    collections::HashMap,
+    fmt::{self, Debug, Formatter},
+};
+
+use arrow_array::RecordBatch;
+use arrow_schema::{Field as ArrowField, Schema as ArrowSchema};
+
+use super::field::Field;
+use crate::arrow::*;
+use crate::{format::pb, io::object_reader::ObjectReader, Error, Result};
+
+/// Lance Schema.
+#[derive(Default, Debug, PartialEq, Clone)]
+pub struct Schema {
+    /// Top-level fields in the dataset.
+    pub fields: Vec<Field>,
+    /// Metadata of the schema
+    pub metadata: HashMap<String, String>,
+}
+
+impl Schema {
+    /// Project the columns over the schema.
+    ///
+    /// ```ignore
+    /// let schema = Schema::from(...);
+    /// let projected = schema.project(&["col1", "col2.sub_col3.field4"])?;
+    /// ```
+    pub fn project<T: AsRef<str>>(&self, columns: &[T]) -> Result<Self> {
+        let mut candidates: Vec<Field> = vec![];
+        for col in columns {
+            let split = col.as_ref().split('.').collect::<Vec<_>>();
+            let first = split[0];
+            if let Some(field) = self.field(first) {
+                let projected_field = field.project(&split[1..])?;
+                if let Some(candidate_field) = candidates.iter_mut().find(|f| f.name == first) {
+                    candidate_field.merge(&projected_field)?;
+                } else {
+                    candidates.push(projected_field)
+                }
+            } else {
+                return Err(Error::Schema(format!(
+                    "Column {} does not exist",
+                    col.as_ref()
+                )));
+            }
+        }
+
+        Ok(Self {
+            fields: candidates,
+            metadata: self.metadata.clone(),
+        })
+    }
+
+    /// Intersection between two [`Schema`].
+    pub fn intersection(&self, other: &Self) -> Result<Self> {
```

Comment:
This is what is added in this PR.

Reply:
so intersection only happens at the top level then? do we need to also look at child fields?

Reply:
Yes. Only these.